### ISSUE TITLE
Question on pytorch imagenet tutorial

### ISSUE THREAD

**Issue by zyainfal** [NVIDIA/DALI#1615]
Hi， I follow the https://github.com/NVIDIA/DALI/blob/master/docs/examples/pytorch/resnet50/main.py to use dali as my dataloader but I encounter this warning:
``
...
Epoch: [7 | 120]
(5004/5004) Data: 0.660s | Batch: 1.042s | loss: 2.3802 | top1:  47.0695 | top5:  71.1468
WARNING:root:DALI iterator does not support resetting while epoch is not finished. Ignoring...
learning rate set to: 0.049581372689098874

Epoch: [8 | 120]
Train
(5005/5004) Data: 0.668s | Batch: 1.049s | loss: 2.3039 | top1:  48.5538 | top5:  72.4979
WARNING:root:DALI iterator does not support resetting while epoch is not finished. Ignoring...
...
``

As I refer to https://github.com/NVIDIA/DALI/blob/master/dali/python/nvidia/dali/plugin/pytorch.py. If I use reset() before an epoch ends, reset() would not work which would lead the iteration to 
``
if self._counter >= self._size:
        if self._auto_reset:
            self.reset()
        raise StopIteration
``
and raise an error. 
However in my case, I can train it many epoches, is there any problems here?

**Comment by jantonguirao** [comment]
Hi @zyainfal. The call to reset should happen at the end of the epoch. This warning message indicates that the call to reset had happened before the end of the epoch and will be ignored. It is hard to say if there is any problem without looking at the code. Could you please share the actual code snippet that you used?

**Comment by zyainfal** [comment]
Thanks for your reply, I initialize the dataloader like this

    pipe = HybridTrainPipe(batch_size=args.batch_size, num_threads=args.workers, 
    device_id=args.local_rank, data_dir=traindir, crop=crop_size, dali_cpu=args.dali_cpu)
    pipe.build()
    train_loader = DALIClassificationIterator(pipe, size=int(pipe.epoch_size("Reader") / args.world_size))

    pipe = HybridValPipe(batch_size=args.batch_size, num_threads=args.workers, 
    device_id=args.local_rank, data_dir=valdir, crop=crop_size, size=val_size)
    pipe.build()
    val_loader = DALIClassificationIterator(pipe, size=int(pipe.epoch_size("Reader") / args.world_size))

And the training code is 

    best_acc = 0.
    log_writer = SummaryWriter("log/imagenet/" + name + "/")
    for epoch in range(args.epoch):
        adjust_lr(optimizer, epoch, args)
        print('\nEpoch: [%d | %d]' % (epoch + 1, args.epoch))
        print("Train")
        train_acc = train(train_loader, model, criterion, optimizer, epoch, args, use_cuda, log_writer)
        print('')
        if (epoch+1) % 10 == 0 or epoch>args.epoch-10 or epoch==0:
            print("Validating")
            test_acc1, test_acc5 = test(val_loader, mdl, criterion, epoch, use_cuda)
            print("")
            print("")
            if test_acc1 >= best_acc:
                model_save(model.cpu(), args.save_folder + name + ".pth")
                model.cuda()
                best_acc = test_acc1
                print("Model saved on acc: {}".format(best_acc))
            print("Current acc: {}/{}".format(test_acc1, best_acc))
            log_writer.add_scalar('test_acc/top1', test_acc1, epoch)
            log_writer.add_scalar('test_acc/top5', test_acc5, epoch)
        train_loader.reset()
        val_loader.reset()
    model_save(model.cpu(), args.save_folder + name + "_with_top1_{}-{}.pth".format(test_acc1, best_acc))
    log_writer.close()
 The train() function is here

    def train(train_loader, model, criterion, optimizer, epoch, args, use_cuda, writer):
    model.train()

    batch_time = AverageMeter()
    data_time = AverageMeter()
    losses = AverageMeter()
    top1 = AverageMeter()
    top5 = AverageMeter()
    end = time.time()
    train_loader_len = int(train_loader._size / args.batch_size)
    
    for batch_idx, data in enumerate(train_loader):
        inputs = data[0]["data"]
        targets = data[0]["label"].squeeze().cuda().long()
        data_time.update(time.time() - end)

        if use_cuda:
            inputs, targets = inputs.cuda(), targets.cuda(async=True)

        # compute output
        outputs = model(inputs)
        loss = criterion(outputs, targets)

        # compute gradient and do SGD step
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        # measure elapsed time
        batch_time.update(time.time() - end)
        end = time.time()

        # measure accuracy and record loss
        prec1, prec5 = accuracy(outputs.data, targets.data, topk=(1, 5))
        top1.update(prec1.item(), inputs.size(0))
        top5.update(prec5.item(), inputs.size(0))   

        loss = loss.float()
        losses.update(loss.data.item(), inputs.size(0))
        print(
                '\r({batch}/{size}) Data: {data:.3f}s | Batch: {bt:.3f}s | Loss: {loss:.4f} | top1: {top1: .4f} | top5: {top5: .4f}'.format(
                    batch=batch_idx + 1,
                    size=train_loader_len,
                    data=data_time.avg,
                    bt=batch_time.avg,
                    loss=losses.avg,
                    top1=top1.avg,
                    top5=top5.avg,
                ), end="")
            if((batch_idx+1)%1001==0):
                writer.add_scalar('losses/loss', losses.avg, batch_idx + epoch*train_loader_len)
                writer.add_scalar('train_acc/top1', top1.avg, batch_idx + epoch*train_loader_len)
                writer.add_scalar('train_acc/top5', top5.avg, batch_idx + epoch*train_loader_len)
                losses.reset()
                top1.reset()
                top1.reset()
    return top1.avg

which is similar to the tutorial. And I noticed that this happens in the official MXNet document 
https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/mxnet/mxnet-resnet50.html like this

INFO:root:Epoch[0] Batch [100]  Speed: 4407.30 samples/sec      accuracy=0.001141
INFO:root:Epoch[0] Batch [200]  Speed: 4444.77 samples/sec      accuracy=0.003184
INFO:root:Epoch[0] Batch [300]  Speed: 4395.88 samples/sec      accuracy=0.006074
INFO:root:Epoch[0] Batch [400]  Speed: 4384.70 samples/sec      accuracy=0.011182
INFO:root:Epoch[0] Batch [500]  Speed: 4389.42 samples/sec      accuracy=0.017441
INFO:root:Epoch[0] Batch [600]  Speed: 4382.10 samples/sec      accuracy=0.026377
INFO:root:Epoch[0] Batch [700]  Speed: 4388.26 samples/sec      accuracy=0.036611
INFO:root:Epoch[0] Batch [800]  Speed: 4383.51 samples/sec      accuracy=0.047139
INFO:root:Epoch[0] Batch [900]  Speed: 4402.73 samples/sec      accuracy=0.057686
INFO:root:Epoch[0] Batch [1000] Speed: 4392.32 samples/sec      accuracy=0.067861
INFO:root:Epoch[0] Batch [1100] Speed: 4384.42 samples/sec      accuracy=0.079248
INFO:root:Epoch[0] Batch [1200] Speed: 4385.37 samples/sec      accuracy=0.090088
INFO:root:Epoch[0] Train-accuracy=0.098537
INFO:root:Epoch[0] Time cost=295.153
**_WARNING:root:DALI iterator does not support resetting while epoch is not finished. Ignoring..._**
INFO:root:Epoch[0] Validation-accuracy=0.104393

would this reminds you what would happen in this case?
Thanks again.

**Comment by JanuszL** [comment]
In your case it comes from:
```
val_loader.reset()
```
You reset it every epoch, but only consume data from it:
```
if (epoch+1) % 10 == 0 or epoch>args.epoch-10 or epoch==0:
```
You should move this reset to the if above it.

**Comment by zyainfal** [comment]
Yes, ur right. Thanks. 
And by the way, this implement is much faster than pytorch ImageFolder, great job!

**Comment by AlbertiPot** [comment]
Reset in the middle of the epoch 

Hi, the DALI dataloader is really much faster! However, I encountered some unusual cases: In my model, I have to explore only part of the batches from an epoch to calibrate the BN layer. Once finished calibrating, the dataloader should be reset. But the reset() function currently only supports resetting at the end of an epoch. Is it any way to reset in the middle of the epoch? Many many thanks
